### DISCUSSION TITLE
How does CSR and SSR/SSG work together

### DISCUSSION THREAD

**Discussion by mo-alaa** [vercel/next.js#55958] [Help] [answered]
### Summary

I wonder how CSR and SSR work together... If I have a page that has 1 component that needs to be available only on client side, however the rest of the page can be available on the server. 

Does this actually happen, that only the part tied to the client runs on the client and the rest run on the server.

Take a look at the following code example of the root _app.js:
```
import { ApolloProvider } from '@apollo/client';
import { useApollo } from '../useApollo';
import { useEffect, useState } from 'react';

function MyApp({ Component, pageProps }) {
  const [isMounted, setIsMounted] = useState(false);
  useEffect(() => {
    setIsMounted(true);
  }, []);

  const client = useApollo(pageProps);

  return (
    <ApolloProvider client={client}>
      {isMounted && <Component1 />}

      <Component {...pageProps} />
    </ApolloProvider>
  );
}

export default MyApp;
``` 

Does **Component1** get rendered on client and the rest of the pages/components rendered on server as it's the default behavior in nextjs ? Or using the `useEffect` has blocked the whole pages from running SSR/SSG ?

**Comment by icyJoseph** [comment]
Hi,

In this case, you got it right. `Component1` won't be rendered during SSR. 

More precisely, `Component1`, is not rendered on the server, and on the client it is not rendered, until after hydration is done. So once React is done with all of its work of rendering the first UI. Then it looks for effects, and there you call `setIsMounted` which triggers a new React render, and then `Component1` is finally touched.

**Reply by mo-alaa** [reply]
But the rest of the pages will be rendered (by default) on the server, am I right ?

**Reply by icyJoseph** [reply, marked as answer]
Right, everything that the SSR pass can touch. In general, use-effect triggered stuff has to be wait to the client runtime. Interactivity triggered stuff much of the same, but you wait for user interactions instead.

The tricky thing is that if you have for example an `isLoading` flag on a provider a high level in your App tree, then you could potentially, by mistake, skip SSR:

```jsx

if (isLoading) return null;

return <>{children}</>
```

If the `isLoading` resolves only client side, then you'll do no work during SSR. So SEO bots, crawlers and non-JS entities, will miss out on the content.

**Reply by mo-alaa** [reply]
Gotcha, so we shouldn't block the ssr with any client-side checks.

**Reply by mo-alaa** [reply]
Thank you :)
